// inngest/deno/fresh.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serve = exports.name = void 0;
const InngestCommHandler_1 = require("../components/InngestCommHandler");
const consts_1 = require("../helpers/consts");
exports.name = "deno/fresh";
/**
 * With Deno's Fresh framework, serve and register any declared functions with
 * Inngest, making them available to be triggered by events.
 *
 * @public
 */
const serve = (nameOrInngest, fns, opts) => {
    const handler = new InngestCommHandler_1.InngestCommHandler(exports.name, nameOrInngest, fns, opts, (req, env) => {
        const url = new URL(req.url, `https://${req.headers.get("host") || ""}`);
        return {
            url,
            env,
            register: () => {
                if (req.method === "PUT") {
                    return {
                        deployId: url.searchParams.get(consts_1.queryKeys.DeployId),
                    };
                }
            },
            run: async () => {
                if (req.method === "POST") {
                    return {
                        data: (await req.json()),
                        fnId: url.searchParams.get(consts_1.queryKeys.FnId),
                        stepId: url.searchParams.get(consts_1.queryKeys.StepId),
                        signature: req.headers.get(consts_1.headerKeys.Signature) || undefined,
                    };
                }
            },
            view: () => {
                if (req.method === "GET") {
                    return {
                        isIntrospection: url.searchParams.has(consts_1.queryKeys.Introspect),
                    };
                }
            },
        };
    }, ({ body, status, headers }) => {
        return new Response(body, { status, headers });
    }).createHandler();
    return (req) => handler(req, Deno.env.toObject());
};
exports.serve = serve;
